src/decision/src/VisionDecision.cpp

```diff
@@ -24,13 +30,321 @@ VisionDecision::VisionDecision(int argc, char **argv, std::string node_name) {
+ */
+int VisionDecision::getMiddle(int startingPos, int row, bool rightSide, const sensor_msgs::Image::ConstPtr& image_scan){
+
+    int incrementer;
+    int startPixel, endPixel;
+
+    // Depending on chosen side, determine where to start
+    // and how to iterate.
+    if(rightSide){
+        incrementer = -1;
+    }else{
+        incrementer = 1;
+    }
+
+    // Find first pixel of the white line in a certain row.
+    startPixel = VisionDecision::getStartPixel(startingPos, incrementer, row, image_scan);
+
+    // Find last pixel of the white line in a certain row.
+    endPixel = VisionDecision::getEndPixel(startPixel, incrementer, row, image_scan);
+
+    // Return average of the two pixels.
+    return (startPixel + endPixel) / 2;
+}
+
+/**
+ * Returns the white pixel right after the black space in between
+ * the line and the left side or right side of the screen.
+ *
+ * @param startingPos column to start parsing
+ * @param NOISEMAX how large noise can be
+ * @param incrementer decides whether to parse from the left or from the right
+ * @param row determines the row to parse
+ * @param image_scan the image to parse
+ *
+ * @returns the white pixel's column position, null if none found
+ */
+int VisionDecision::getStartPixel(int startingPos, int incrementer, int row,
+                     const sensor_msgs::Image::ConstPtr& image_scan){
+    int column = startingPos;
+    int whiteVerificationCount = 0;
+    int blackVerificationCount = 0;
+    int startPixel = null;
+    int toBeChecked = null;
+
+    // Find starting pixel
+    while(column < image_scan->width && column >= 0){
+        // If white pixel found start verifying if proper start.
+        if(image_scan->data[row*image_scan->width + column] != 0){
+            blackVerificationCount = 0;
+            // This pixel is what we are checking
+            if (toBeChecked == null)
+                toBeChecked = column;
+
+            // Determine whether toBeChecked is noise
+            whiteVerificationCount++;
+            if (whiteVerificationCount == NOISEMAX && startPixel == null)
+                startPixel = toBeChecked;
+        } else {
+            blackVerificationCount++;
+            if(blackVerificationCount == NOISEMAX) {
+                whiteVerificationCount = 0; // Reset verification if black pixel.
+                toBeChecked = null;
+            }
+        }
+        // Go to next element
+        column += incrementer;
+    }
+    return startPixel;
+}
+
+/**
+ * Returns the white pixel right before the black space in between
+ * the lines
+ *
+ * @param startingPos column to start parsing
+ * @param NOISEMAX how large noise can be
+ * @param incrementer decides whether to parse from the left or from the right
+ * @param row determines the row to parse
+ * @param image_scan the image to parse
+ *
+ * @returns the white pixel's column position, null if none found
+ */
+int VisionDecision::getEndPixel(int startingPos, int incrementer, int row,
```

Comment:
`getStartPixel` and `getEndPixel` seem to do fairly similar things? Perhaps much of the functionality of them could be extracted to another function, and the overall code length shortened?